### PR TITLE
Rewrote about dialog, now holds git commit and date.

### PR DESCRIPTION
**Pull request by CandyFace**
The about dialog has been rewritten in c++ instead of html to be able to use variables and other useful things, relevant for developers.
You can now see commit and date from which the version is based on.

The old logo has also been replaced.

screen shot 2017-04-14 at 09 24 26

### CODE REVIEW

**Comment by mapreri** [comment]
Haven't tried, but does that cope if the code is not in a git repository?

In that case I think the box should not even be present.

**Comment by CandyFace** [comment]
Hmm yes that's true, probably not as it calls git from command line.
The problem is though, if we remove it in those cases, then how will we know how old their build is?

**Comment by mapreri** [comment]
Right, in that case you won't have that information that easily.

In case we start realising tarballs we could consider adding a file with the git information about the commit the tarball is based off.

In my use case (the Debian package) the commit date and sha1 are part of the package version, and I could define those variables manually anyway.

**Comment by CandyFace** [comment]
while untested, I've now added a condition which checks if git exists, if it doesn't it should ignore the code and not add the box.

File: app/app.pro [hunk 1]
```diff
@@ -90,9 +93,21 @@ FORMS += \
...
 DEFINES += APP_VERSION=\\\"$$VERSION\\\"
 
+win32 {
+    GIT_EXISTS='"\\\"$$system(git --version > nul)\\\""'
+} else {
+    GIT_EXISTS='"\\\"$$system(git --version > /dev/null)\\\""'
+}
+
+GIT_EXISTS {
+    DEFINES += GIT_CURRENT_SHA1="\\\"$$system(git --git-dir $$_PRO_FILE_PWD_/../.git --work-tree $$_PRO_FILE_PWD_/../ rev-parse HEAD)\\\"" \
```
Comment (J5lx) [inline app/app.pro]: Command could be shortened by using `-C`, see docs at <https://git-scm.com/docs/git#git--Cltpathgt> for example.

**Comment by chchwy** [comment]
Compiling failed on Windows, very weird error message...

```
4>RC : fatal error RC1107: invalid usage; use RC /? for Help
4>
4>         The command exited with code 1.
4>       Done executing task "RC" -- FAILED.
4>     1>
4>Done building target "ResourceCompile" in project "Pencil2D.vcxproj" -- FAILED.
```

**Comment by CandyFace** [comment]
Hmm no idea about that error, however I discovered that the if statement doesn't work correctly. I'll look into that soon. It's possible that the problem is related to that compiler error.

**Comment by CandyFace** [comment]
@chchwy I've made some changes, try to see if it compiles now on Windows

**Comment by chchwy** [comment]
Good job, it compiled.

**Comment by CandyFace** [comment]
Baah.. I can't make appveyor compile it, getting  this error:
```C++
..\..\app\aboutdialog.cpp(53): error C2146: syntax error : missing ')' before identifier 'Tue'
```

I have no idea how to fix it as it works fine on Linux and Mac. Ugh Visual Studio...

**Comment by J5lx** [comment]
I think it’s because GIT_TIMESTAMP is escaped twice: `… /D GIT_EXISTS /D "GIT_CURRENT_SHA1=\"dbc6bda372472340f26fea72a98f5732f64eb2c2\"" /D "GIT_TIMESTAMP=\"\"Tue Apr 18 11:57:10 2017\"\"" …`. So the compiler sees `"… <br> date: " ""Tue Apr 18 11:57:10 2017""));` which is obviously wrong. I think it’s because GIT_TIMESTAMP is wrapped in single quotes in app.pro rather than double quotes. I can’t test right now, though.

**Comment by CandyFace** [comment]
Ahh yes you may be onto something! I'll try to change the single quotes to double and see what happens.

**Comment by J5lx** [comment]
Actually I think the single quotes might be entirely dispensable since there are is already a pair of enclosed, non-escaped double quotes anyway (i.e. `'"\\\"…\\\""'` will become a define value of `"\"…\""` while `"\\\"…\\\""` will become `\"…\"`).

**Comment by CandyFace** [comment]
Right, I thought that would work too, but now it just throws more errors.
 
```c++
GIT_TIMESTAMP="\\\"$$system(git --git-dir=.git --work-tree=. -C $$_PRO_FILE_PWD_/../ log -1 --format=%cd --date=local)\\\""
```

```
"C:\projects\pencil\build\pencil.sln" (Build target) (1) ->
"C:/projects/pencil/build/app\Pencil2D.vcxproj.metaproj" (default target) (7) ->
"C:\projects\pencil\build\app\Pencil2D.vcxproj" (default target) (8) ->
(ClCompile target) -> 
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_aboutdialog.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_actioncommands.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_colorbox.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_colorgrid.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_colorgriditem.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_colorinspector.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_colorpalettewidget.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_colorwheel.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_displayoptionwidget.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_errordialog.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_exportimageseqdialog.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_exportmoviedialog.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_filedialogex.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_imageseqdialog.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_mainwindow2.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_pencilapplication.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_preferencesdialog.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_preview.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_shortcutfilter.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_shortcutspage.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
  c1xx : fatal error C1083: Cannot open source file: 'release\moc_timeline2.cpp': No such file or directory [C:\projects\pencil\build\app\Pencil2D.vcxproj]
    26 Warning(s)
    21 Error(s)

```

**Comment by J5lx** [comment]
I think I got it. The actual source of the error lies within the moc invocation, e.g. `moc: Too many input files specified: 'Apr' '18' '15:44:13' '2017"' '..\..\app\aboutdialog.h'`. The command line for that is this (on my local dev VM anyway):

```
C:\Qt\5.8\msvc2013_64\bin\moc.exe  -DUNICODE -DWIN32 -DWIN64 -DQUAZIP_BUILD -DQUAZIP_STATIC -DAPP_VERSION=\"0.5.4\" -DGIT_EXISTS -DGIT_CURRENT_SHA1=\"03ec5cacfdd188d159321f2ca2c64993cebd6d66\" -DGIT_TIMESTAMP=\"Tue Apr 18 15:44:13 2017\" -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_MULTIMEDIA_LIB -DQT_GUI_LIB -DQT_XML_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB --compiler-flavor=msvc --include debug/moc_predefs.h -IC:/Qt/5.8/msvc2013_64/mkspecs/win32-msvc2013 -IC:/Users/j5lx/Documents/pencil/app -IC:/Users/j5lx/Documents/pencil/core_lib/graphics -IC:/Users/j5lx/Documents/pencil/core_lib/graphics/bitmap -IC:/Users/j5lx/Documents/pencil/core_lib/graphics/vector -IC:/Users/j5lx/Documents/pencil/core_lib/interface -IC:/Users/j5lx/Documents/pencil/core_lib/structure -IC:/Users/j5lx/Documents/pencil/core_lib/tool -IC:/Users/j5lx/Documents/pencil/core_lib/util -IC:/Users/j5lx/Documents/pencil/core_lib/ui -IC:/Users/j5lx/Documents/pencil/core_lib/managers -IC:/Users/j5lx/Documents/pencil/core_lib -IC:/Users/j5lx/Documents/pencil/3rdlib/quazip -IC:/Users/j5lx/Documents/pencil/3rdlib/zlib -IC:/Qt/5.8/msvc2013_64/include -IC:/Qt/5.8/msvc2013_64/include/QtSvg -IC:/Qt/5.8/msvc2013_64/include/QtWidgets -IC:/Qt/5.8/msvc2013_64/include/QtMultimedia -IC:/Qt/5.8/msvc2013_64/include/QtGui -IC:/Qt/5.8/msvc2013_64/include/QtANGLE -IC:/Qt/5.8/msvc2013_64/include/QtXml -IC:/Qt/5.8/msvc2013_64/include/QtNetwork -IC:/Qt/5.8/msvc2013_64/include/QtCore -I. -I"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE" -I"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\ATLMFC\INCLUDE" -I"C:\Program Files (x86)\Windows Kits\8.1\include\shared" -I"C:\Program Files (x86)\Windows Kits\8.1\include\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\winrt" ..\..\app\aboutdialog.h -o debug\moc_aboutdialog.cpp
```

The problem is this: Windose’s cmd uses a circumflex ^ rather than a backslash \\ as escape character. We didn’t get errors before because none of the other defines contain spaces, but since GIT_TIMESTAMP does it triggers that error.

**Comment by CandyFace** [comment]
Interesting, so how do we fix it then?

**Comment by J5lx** [comment]
That’s what I’m trying to figure out next.

**Comment by J5lx** [comment]
I’m going to call it a day for now. I’m starting to get the feeling that I’m actually trying to work around some bug in qmake, since fixing moc breaks compilation and fixing compilation breaks moc, like switching between the two kinds of errors above, no matter how exactly I do it. It feels as if qmake handles compiler args and moc args differently somehow, or maybe it’s somehow related to moc’s --and cl’s /different:syntaxes.

**Comment by CandyFace** [comment]
Yup I agree... I've tried numerous things too, nothing has solved the problem :/

**Comment by CandyFace** [comment]
AHA! I got it, and it's totally not madness at all!  (￣∇ ￣) 
When I finally figure a combination that VS accepts, then travis starts complaining. In the end making two DEFINES seems to work, one for windows and one for the rest...

**Pull request merged**